willow/svg.py

```diff
@@ -0,0 +1,160 @@
+        self.width = self._get_width()
+        self.height = self._get_height()
+
+    @property
+    def root(self):
+        return self.dom.getroot()
+
+    def _get_width(self):
+        if attr := (self.root.get("width") or self.root.get("height")):
+            return self._parse_size(attr)
+        elif self.view_box is not None:
+            return self.view_box.width
+
+    def _get_height(self):
+        if attr := (self.root.get("height") or self.root.get("width")):
+            return self._parse_size(attr)
+        elif self.view_box is not None:
+            return self.view_box.height
+
+    def _parse_size(self, raw_value):
+        clean_value = raw_value.strip()
+        match = self.UNIT_RE.search(clean_value)
+        unit = clean_value[match.start() :] if match else None
+
+        if unit == "%":
+            raise InvalidSizeAttribute(
+                f"Unable to handle relative size units ({raw_value})"
+            )
+
+        # TODO: better handling of invalid, not %, units
+
+        amount_raw = clean_value[: -len(unit)] if unit else clean_value
+        try:
+            # Might be real number or int, may have a sign, so try to
+            # cast to a float rather than (e.g.) using a complex regex.
+            # https://www.w3.org/TR/2008/REC-CSS2-20080411/syndata.html#length-units
+            amount = float(amount_raw)
+        except ValueError as err:
+            raise InvalidSizeAttribute(
+                f"Unable to parse value from '{raw_value}'"
+            ) from err
+
+        if unit is None or unit == "px":
+            return round(amount)
+        elif unit == "em":
+            return round(amount * self.font_size_px)
+        elif unit == "ex":
+            # This is not exactly correct, but it's the best we can do
+            return round(amount * self.font_size_px / 2)
+        else:
+            return round(amount * self.dpi * self.COEFFICIENTS[unit])
+
+    def _get_view_box(self):
+        if attr := self.root.get("viewBox"):
+            return self._parse_view_box(attr)
+
+    def _parse_view_box(self, raw_value):
+        match = self.VIEW_BOX_RE.match(raw_value.strip())
+        if match is None:
+            raise ViewBoxParseError(f"Unable to parse viewBox value '{raw_value}'")
+        return ViewBox(*map(int, match.groups()))
+
+    def write(self, f):
+        self.dom.write(f, encoding="utf-8")
+        f.seek(0)
+
+
+class SVGImageOperation(str, Enum):
+    CROP = "crop"
+    RESIZE = "resize"
+
+
+class SVGImage(Image):
+    def __init__(self, dom, operations=None, dpi=96, font_size="12pt"):
+        self.image = SVGWrapper(dom)
+        self.operations = [] if operations is None else operations
+
+    @classmethod
+    @Image.converter_from(SVGImageFile)
+    def open(cls, svg_image_file):
+        return cls(svg_image_file.dom)
+
+    @Image.operation
+    def get_size(self):
+        return (self.image.width, self.image.height)
```

Comment:
The size needs to change when operations are applied on the image. Looks like this will always return the size of the original image.

Reply:
It looks like we may have no choice but to rasterise to PNG at this point, and apply any outstanding operations to get the latest width/height. Perhaps we can cache the PNG version on the current instance, so that it doesn't go to waste?

Reply:
We can figure out the new size as we go without having to rasterise. The only operations willow supports that alter the size of the image are crop and resize and both of those will tell us the size of the image after the operation.

Reply:
Okie dokie. In that case, is it worth us applying the resize/crop operations on the SVG lazily? It sounds like it would be common enough to need the resulting width/height that we might as well apply SVG transformations as we go?

Reply:
I've changed the way this is handled somewhat - crops and resizes can be applied directly to SVGs, with a couple of small assumptions. I'll add more detail in a top-level comment